playbooks/continuous_delivery/ci-cd-elements.adoc

```diff
@@ -0,0 +1,252 @@
+|https://en.wikipedia.org/wiki/Chaos_Monkey[Chaos test]|https://www.gremlin.com/[Gremlin], https://fabric8.io/guide/chaosMonkey.html[Fabric8 Chaos Monkey].
+|===
+
+Behavioral Driven Development (BDD) testing is a valuable option and should be used for testing web and mobile applications. +
+BDD test cases are created using a pseudo natural language (https://cucumber.io/docs/reference[Gerkin]), allowing them to be quickly and easily created by non-technical users. The test cases are translated into machine instructions at test runtime (a popular one is https://cucumber.io/[Cucumber]).  +
+The general architecture of BDD tests is illustrated in the following diagram:
+
+image::images/BDD.png[]
+
+The above architecture can be fully deployed and contained in OpenShift. Multiple combinations of browser/operating system and mobile device/operating system tests can be executed in parallel. The advent of Windows containers will provide additional opportunities for testing applications running on Internet Explorer and Edge.
+
+=== Compliance steps
+
+Compliance steps may be needed when pipelines must be compliant with an organization's release process and change management.
+
+Examples of compliance steps are provided in the following table:
+
+|===
+|*Step *|*Description*
+
+|Manual approval|Suspends pipeline execution until manual  approval to continue is provided .
+|Container image signing |Cryptographically signs images 
+|Deployment manifest upload|Uploads the manifest of what is deployed to a centralized change management tool.
+|===
+
+Container image signing provides a method for making an assertion cryptographically verifiable. When an image is signed (a fact to which any meaning can be assigned, for example, the image is approved for execution ), anyone can verify that assertion. Again having assertions is only one piece of the process, the other piece, which resides outside of the pipeline, is having a way to verify and enforce them. https://github.com/redhat-cop/openshift-image-signing-scanning[Here] you can find an example of a pipeline with an container image signing step.
+
+Google is currently working on a more general purpose assertion framework for deployment pipelines (see https://cloudplatform.googleblog.com/2017/10/introducing-grafeas-open-source-api-.html[Grapheas and Kritis]).
+
+== Pipeline metrics and dashboards
+
+A wealth of metrics can be collected from CI/CD pipelines, and much of the value of a pipeline is lost if metrics are not collected and made available in dashboards. DevOps teams utilize metrics for fast feedback.
+
+Useful metrics might include: number of deployment errors  and total pipeline execution time. 
+
+Although there are not many tools currently available for providing pipeline metrics, one example worth examining is https://github.com/capitalone/Hygieia[Hygieia] from https://developer.capitalone.com/open-source/[Capital One].
+
+A pipeline dashboard is illustrated in the following screen capture:
```

Comment:
There is no image or illustration following this statement displaying a pipeline dashboard.  Please add or remove statement.

Reply:
added missing image